src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStreamAsyncMemoryOverloads.cs

```diff
@@ -0,0 +1,131 @@
+                                                                                        s_localizableTitleRead,
+                                                                                        s_localizableMessageRead,
+                                                                                        DiagnosticCategory.Performance,
+                                                                                        RuleLevel.IdeSuggestion,
+                                                                                        s_localizableDescriptionRead,
+                                                                                        isPortedFxCopRule: false,
+                                                                                        isDataflowRule: false);
+
+        internal static DiagnosticDescriptor PreferStreamWriteAsyncMemoryOverloadsRule = DiagnosticDescriptorHelper.Create(
+                                                                                        RuleId,
+                                                                                        s_localizableTitleWrite,
+                                                                                        s_localizableMessageWrite,
+                                                                                        DiagnosticCategory.Performance,
+                                                                                        RuleLevel.IdeSuggestion,
+                                                                                        s_localizableDescriptionWrite,
+                                                                                        isPortedFxCopRule: false,
+                                                                                        isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+            ImmutableArray.Create(PreferStreamReadAsyncMemoryOverloadsRule, PreferStreamWriteAsyncMemoryOverloadsRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterOperationAction(OnCompilationStart, OperationKind.Invocation);
+        }
+
+        private static void OnCompilationStart(OperationAnalysisContext context)
+        {
+            if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemIOStream, out INamedTypeSymbol? stream) ||
+                stream == null)
+            {
+                return;
+            }
+
+            // Verify that the required Memory types were shipped in the current .NET version, otherwise we should not suggest the overload
+            if (context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemMemoryExtensions) == null ||
```

Comment:
Since I only want to know if the type exists, is this the most efficient way we have to do such check?

Reply:
Why do you need MemoryExtensions?  The ReadAsync/WriteAsync methods are declared directly on Stream.

I'd expect:

* Find `System.Memory<T>`
* Find System.IO.Stream
* Check for the presence of the correct WriteAsync overload on Stream (something like https://github.com/dotnet/roslyn-analyzers/blob/0cbdd68e14722e911d50eee64ee95823b3ffa09c/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotUseReferenceEqualsWithValueTypes.cs#L59-L72, though you'll need to either do generic instantiation to look forward, or loop over each one and look for `memoryType.Equals(parameter.Type.OriginalDefinition)`.

That's looking for the thing you actually are evaluating for, as opposed to a convenient nearby marker.

I don't know how symbols and overloads work, but it's possible that binding the -other- overloads while looking for the desired target method will allow for a fast "am I invoking one of the overloads I'm reporting against?" test.

Reply:
The _old_ ReadAsync/WriteAsync methods are declared directly on Stream, yes, but the new overloads, the ones we want to suggest, are declared in MemoryExtensions.

There's no point in suggesting to the user to change the ReadAsync/WriteAsync overloads if they are using a version of .NET Core that is too old (< 2.1).

You're right, I don't need to check for all 3 types to exist, I only need to check for one: either SystemMemory or SystemReadOnlyMemory.

Reply:
> but the new overloads, the ones we want to suggest, are declared in MemoryExtensions

This is not true. They're all on Stream directly.

Reply:
Oh, you're right! Thanks for correcting me. I don't know why I was confusing methods, sorry about that.

What I meant to say was that the `AsMemory` methods are on `MemoryExtensions`. I'm using that method to retrieve the `Memory` object from the byte array in my unit tests and pass it to the Read/WriteAsync method.

I found the `AsMemory` extension methods were introduced in [.NET Standard 2.1](https://github.com/dotnet/dotnet-api-docs/blob/31a50e14c17ae7713232305087b00558dc522905/xml/System/MemoryExtensions.xml#L47). I guess it's expected that the new WriteAsync/ReadAsync methods were also introduced in [.NET Standard 2.1](https://github.com/dotnet/dotnet-api-docs/blob/master/xml/System.IO/Stream.xml#L2700). 

.NET Standard is not among the versions that [ReferenceAssemblies.Default can return](https://github.com/dotnet/roslyn-sdk/blob/ea2ad80d27b049eaca163044c6909114930224e9/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/ReferenceAssemblies.cs#L85-L101), hence the trick suggested by Jeremy to specify a particular .NET Core version to run the tests, which worked.

Reply:
In my latest commit, I try to see if `System.Memory` exists in the current reference assemblies. If it is not found, we quit early. See my latest commits. I'm not looking for `MemoryExtensions` anymore.

If I didn't search for `System.Memory`, I would have to find out if the new `ReadAsync/WriteAsync` methods are available in `Stream`. That would mean going through a lot more code than necessary.

Reply:
You don't need AsMemory or MemoryExtensions.  You can just construct a memory directly, e.g. `new Memory<byte>(array, offset, count)` instead of `array.AsMemory(offset, count)`... the latter just calls the former.

Reply:
That suggestion is for the fixer, right?

Reply:
> That suggestion is for the fixer, right?

Or tests.  You don't need MemoryExtensions.AsMemory anywhere in this PR, is my point :)

Reply:
I added unit tests with both cases: `new Memory<byte>(array, offset, count)` and `array.AsMemory(offset, count)`.